### ISSUE TITLE
Ensure vertical thickness causes occasionally solid infill mid object & doesn't work with 3 perimeter and more

### ISSUE THREAD

**Issue by rovili** [supermerill/SuperSlicer#2254]
### What happened?

Solid infills are created mid object when "ensure vertical thickness" is activated.
It does not happen with the newest and older PrusaSlicer versions.

Super Slicer - solid infill example
<img width="671" alt="image" src="https://user-images.githubusercontent.com/97400836/150569866-2abc745c-ce04-481b-a693-04aebc78b78e.png">

Prusa Slicer - solid infill example
<img width="557" alt="image" src="https://user-images.githubusercontent.com/97400836/150570076-8536322e-ed40-4b65-8a7e-c28ac9df621d.png">

Super Slicer - 3 perimeter thickness/overhang-support:
<img width="528" alt="image" src="https://user-images.githubusercontent.com/97400836/150570568-9133f1cb-8cdb-4628-9ef2-e06e859dfbe3.png">

Prusa Slicer - 3 perimeter thickness/overhang-support:
<img width="557" alt="image" src="https://user-images.githubusercontent.com/97400836/150570834-9f03dc09-83ba-4947-a962-234e11fa752b.png">

### Project file & How to reproduce

[moai-small-no_overhang.3mf.zip](https://github.com/supermerill/SuperSlicer/files/7915460/moai-small-no_overhang.3mf.zip)
[3perimeter_moai-small-no_overhang.3mf.zip](https://github.com/supermerill/SuperSlicer/files/7915496/3perimeter_moai-small-no_overhang.3mf.zip)

### Version

2.3.57.9

### Operating system

macOS 12.2

**Comment by supermerill** [comment]
Can you try without the Horizontal shell ->"no soilid infill over ... perimeters" option?

**Comment by rovili** [comment]
<img width="708" alt="image" src="https://user-images.githubusercontent.com/97400836/150762301-a7ed8313-76e4-4e29-90fc-e3e2dcdf5d8c.png">

<img width="691" alt="image" src="https://user-images.githubusercontent.com/97400836/150762343-4cb7f244-554b-43a5-9a69-48b47289bf3f.png">

**Comment by supermerill** [comment]
Superslicer with 0 for "no solid infill over...":
![image](https://user-images.githubusercontent.com/6536403/150763393-26839e0e-3c20-4f39-997a-837d0376ccc5.png)
seems the same as ps.
Is it okay for you?

**Comment by rovili** [comment]
You're right, with 0 it helps for the overhang and it is similar/same as in PS. I agree.

But my picture with solid infill from above was from today with 0 - it happens only occasionally.

**Comment by acwest** [comment]
I am getting similar mysterious solid infill layers, in version 2.3.57.10. The `ensure vertical thickness` and `no solid infill over` settings do not seem to affect it, in my case. In the screenshot below, you can see that some layers are marked as having an overhang. At those layer heights, a complete top solid infill is inserted below the layer, and a complete bottom solid infill is inserted above it. Another thing that is interesting is that the central sun gear that is showing these problems does not have the overhangs or extra layers when sliced further from the planet gears. Those are probably two separate issues, but I don't really believe in coincidence...

![Screenshot from 2022-02-09 15-11-54](https://user-images.githubusercontent.com/55695/153286060-175aecbc-0c2f-40d6-8bfe-41b90b41534a.png)
